calico.go

```diff
@@ -306,57 +306,84 @@ func cmdDel(args *skel.CmdArgs) error {
+				// so we have no way of knowing/checking ActiveInstanceID. To protect the endpoint
+				// from false DEL, we return the error without deleting/cleaning up.
+				return err
+			}
 		}
-		logger.WithField("stdin", args.StdinData).Debug("Updated stdin data for Delete Cmd")
-	}
 
-	ipamErr := ipam.ExecDel(conf.IPAM.Type, args.StdinData)
+		// Store WEP Metadata in endpoint only if it is not nil to avoid nil pointer dereference
+		// in some of the following operations when WEP doesn't exist in the datastore.
+		if wep != nil {
+			endpoint = wep.Metadata
+		} else {
+			endpoint = ep
+		}
 
-	if ipamErr != nil {
-		logger.Error(ipamErr)
+		// Check if ActiveInstanceID is populated (it will be an empty string "" if it was populated
+		// before this field was added to the API), and if it is there then compare it with ContainerID
+		// passed by the orchestrator to make sure they are the same, return without deleting if they aren't.
+		if endpoint.ActiveInstanceID != "" && args.ContainerID != endpoint.ActiveInstanceID {
+			logger.WithField("WorkloadEndpoint", wep).Warning("CNI_ContainerID does not match WorkloadEndpoint ActiveInstanceID so ignoring the DELETE cmd.")
+			return nil
+		}
 	}
 
-	calicoClient, err := CreateClient(conf)
-	if err != nil {
-		return err
+	// Store WEP Metadata in endpoint only if it is not nil to avoid nil pointer dereference
+	// in some of the following operations when WEP doesn't exist in the datastore.
+	if wep != nil {
+		endpoint = wep.Metadata
+	} else {
+		endpoint = ep
 	}
 
-	ep := api.WorkloadEndpointMetadata{
-		Name:         args.IfName,
-		Node:         nodename,
-		Orchestrator: orchestrator,
-		Workload:     workload}
-	if err = calicoClient.WorkloadEndpoints().Delete(ep); err != nil {
-		if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
-			logger.WithField("endpoint", ep).Info("Endpoint object does not exist, no need to clean up.")
-		} else {
+	// Delete the WorkloadEndpoint object from the datastore.
+	if err = calicoClient.WorkloadEndpoints().Delete(endpoint); err != nil {
+		switch err := err.(type) {
+		case errors.ErrorResourceDoesNotExist:
+			// Log and proceed with the clean up if WEP doesn't exist.
+			logger.WithField("endpoint", endpoint).Info("Endpoint object does not exist, no need to clean up.")
+		case errors.ErrorResourceUpdateConflict:
+			// This case means the WEP object was modified between the time we did the Get and now,
+			// so it's not a safe Compare-and-Delete operation, so log and abbort with the error.
+			logger.WithField("endpoint", endpoint).Warning("Error deleting endpoint: endpoint was modified before it could be deleted.")
+			return fmt.Errorf("Error deleting endpoint: endpoint was modified before it could be deleted: %v", err)
+		default:
 			return err
 		}
 	}
 
+	// Release the IP address by calling the configured IPAM plugin.
+	ipamErr := cleanUpIPAM(conf, args, logger)
+
 	// Only try to delete the device if a namespace was passed in.
```

Comment:
Since you've pulled cleanUpIpam into its own function, does it make sense to put the namespace cleanup in its own function too?